File: load/repertoire_load.py
#
# Import data into AIRR mongo repository for a public project. This is for
# the repertoire metadata. This assumes you are running in the docker container.
#

import json
from dotenv import load_dotenv
import os
import airr
import yaml
import requests
import argparse

# Setup
def getConfig():
    if load_dotenv(dotenv_path='/api-js-tapis/.env'):
        cfg = {}
        cfg['api_server'] = os.getenv('WSO2_HOST')
        cfg['api_key'] = os.getenv('WSO2_CLIENT_KEY')
        cfg['api_secret'] = os.getenv('WSO2_CLIENT_SECRET')
        cfg['username'] = os.getenv('VDJ_SERVICE_ACCOUNT')
        cfg['password'] = os.getenv('VDJ_SERVICE_ACCOUNT_SECRET')
        return cfg
    else:
        print('ERROR: loading config')
        return None

# Fetches a user token based on the supplied auth object
# and returns the auth object with token data on success
def getToken(config):
    data = {
        "grant_type":"password",
        "scope":"PRODUCTION",
        "username":config['username'],
        "password":config['password']
    }
    headers = {
        "Content-Type":"application/x-www-form-urlencoded"
    }

    url = 'https://' + config['api_server'] + '/token'

    resp = requests.post(url, data=data, headers=headers, auth=(config['api_key'], config['api_secret']))
    token = resp.json()
    return token

# Insert a repertoire by first deleting any repertoire with the same id
# then inserting the new repertoire
def insertRepertoire(token, config, rep):
    headers = {
        "Content-Type":"application/json",
        "Accept": "application/json",
        "Authorization": "Bearer " + token['access_token']
    }

    # delete that repertoire_id
    url = 'https://' + config['api_server'] + '/meta/v3/v1public/repertoire/*?filter=' + requests.utils.quote('{"repertoire_id":"' + rep['repertoire_id'] + '"}')
    print(url)
    resp = requests.delete(url, headers=headers)
    print(resp.json())

    # insert the repertoire
    url = 'https://' + config['api_server'] + '/meta/v3/v1public/repertoire/'
    data = [ rep ]
    resp = requests.post(url, json=data, headers=headers)
    print(resp.json())


# main entry
if (__name__=="__main__"):
    parser = argparse.ArgumentParser(description='Load AIRR repertoire metadata into repository.')
    parser.add_argument('repertoire_file', type=str, help='Repertoire metadata file name')
    args = parser.parse_args()

    if args:
        data = airr.load_repertoire(args.repertoire_file)

        config = getConfig()
        print(config)
        token = getToken(config)
        print(token['access_token'])

        reps = data['Repertoire']

        for r in reps:
            if r.get('repertoire_id') is None:
                print('Repertoire is missing repertoire_id')
                sys.exit(0)
            if len(r['repertoire_id']) == 0:
                print('Repertoire is missing repertoire_id')
                sys.exit(0)
            insertRepertoire(token, config, r)

File: app/load/repertoire_load.js
//
// Import data into AIRR mongo repository for a public project. This is for
// the repertoire metadata. This assumes you are running in the docker container.
//

'use strict';

var csv = require('csv-parser');
var fs = require('fs');
var MongoClient = require('mongodb').MongoClient;
var assert = require('assert');

// Server environment config
var config = require('../config/config');
var mongoSettings = require('../config/mongoSettings');

// Insert a repertoire by first deleting any repertoire with the same id
// then inserting the new repertoire
function insertRepertoire(rep) {
    // get connection to database
    MongoClient.connect(mongoSettings.url, function(err, db) {
        if (err) {
            console.error("Could not connect to database: " + err);
            return;
        } else {
            var v1airr = db.db(mongoSettings.dbname);
            var collection = v1airr.collection('repertoire');

            // delete than insert repertoire
            collection.deleteOne({"repertoire_id":rep['repertoire_id']})
                .then(function(result) {
                    console.log('Deleted repertoire: ' + rep['repertoire_id']);
                    return collection.insertOne(rep);
                })
                .then(function(result) {
                    console.log('Inserted repertoire: ' + rep['repertoire_id']);
                    db.close();
                });
        }
    });
}

console.log('Load AIRR repertoire metadata into repository.');
if (process.argv.length != 3) {
    console.log('usage: node repertoire_load.js repertoire_file');
    process.exit(1);
}

var myArgs = process.argv.slice(2);
console.log(myArgs);
console.log(mongoSettings.url);

var text = fs.readFileSync(myArgs[0]);
var data = JSON.parse(text);
var reps = data['Repertoire'];

for (var i in reps) {
    var r = reps[i];
    if (!r['repertoire_id']) {
        console.error('Repertoire is missing repertoire_id');
        process.exit(1);
    }
    if (r['repertoire_id'].length == 0) {
        console.error('Repertoire is missing repertoire_id');
        process.exit(1);
    }
    insertRepertoire(r);
}
